File: android/app/src/main/java/com/xaymaca/sit/ui/warm/WarmFilterChip.kt
package com.xaymaca.sit.ui.warm

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.wrapContentWidth
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.xaymaca.sit.ui.theme.WarmCategory
import com.xaymaca.sit.ui.theme.Warmth
import com.xaymaca.sit.ui.theme.WarmTheme

sealed class FilterChipKind {
    object All : FilterChipKind()
    data class CategoryKind(val category: WarmCategory) : FilterChipKind()

    /** TIC-88: a user-created group filter chip. Neutral pill styling (like the
     *  All chip) rather than a canonical category's accent, so custom groups read
     *  as distinct from the four canonical categories. */
    object GroupKind : FilterChipKind()
}

/**
 * Pill-shaped filter chip used on the warm Network screen. Mirrors
 * iOS WarmFilterChip — label locks to one line + intrinsic width so it
 * never collapses inside a horizontally scrolling row.
 */
@Composable
fun WarmFilterChip(
    kind: FilterChipKind,
    label: String,
    isActive: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    count: Int? = null,
    warmth: Warmth = Warmth.Subtle,
) {
    val palette = WarmTheme.palette(warmth)

    // GroupKind shares the neutral All-chip styling (ink when active, cardBg
    // otherwise) so custom groups read distinctly from canonical categories.
    val neutralActive = (kind is FilterChipKind.All || kind is FilterChipKind.GroupKind) && isActive
    val neutralInactive = (kind is FilterChipKind.All || kind is FilterChipKind.GroupKind) && !isActive
    val background = when {
        neutralActive -> palette.ink
        neutralInactive -> palette.cardBg
        kind is FilterChipKind.CategoryKind && isActive -> kind.category.palette.accent
        kind is FilterChipKind.CategoryKind && !isActive -> kind.category.palette.accentTint
        else -> palette.cardBg
    }
    val foreground = when {
        neutralActive -> palette.paper
        neutralInactive -> palette.ink
        kind is FilterChipKind.CategoryKind && isActive -> Color(0xFFFAF4E2)
        kind is FilterChipKind.CategoryKind && !isActive -> kind.category.palette.accent
        else -> palette.ink
    }
    val borderColor = when {
        kind is FilterChipKind.CategoryKind && !isActive -> palette.cardBorder
        kind is FilterChipKind.GroupKind && !isActive -> palette.cardBorder
        else -> Color.Transparent
    }

    val countBg = when {
        neutralActive -> palette.paper.copy(alpha = 0.18f)
        neutralInactive -> palette.ink.copy(alpha = 0.06f)
        kind is FilterChipKind.CategoryKind && isActive -> Color.White.copy(alpha = 0.22f)
        kind is FilterChipKind.CategoryKind && !isActive -> kind.category.palette.accentBadge
        else -> palette.ink.copy(alpha = 0.06f)
    }

    Row(
        modifier = modifier
            .wrapContentWidth()
            .clip(CircleShape)
            .background(background)
            .border(1.dp, borderColor, CircleShape)
            .clickable(onClick = onClick)
            .padding(horizontal = 14.dp, vertical = 10.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        if (kind is FilterChipKind.CategoryKind) {
            // Size matches the 13sp label glyph height so CategoryKind
            // chips don't render taller than the icon-less All chip.
            Icon(
                imageVector = kind.category.icon,
                contentDescription = null,
                tint = foreground,
                modifier = Modifier.size(16.dp),
            )
        }
        Text(
            text = label,
            style = TextStyle(
                fontSize = 13.sp,
                fontWeight = FontWeight.SemiBold,
                color = foreground,
            ),
            maxLines = 1,
            overflow = TextOverflow.Visible,
        )
        if (count != null) {
            Box(
                modifier = Modifier
                    .clip(CircleShape)
                    .background(countBg)
                    .padding(horizontal = 6.dp, vertical = 2.dp),
            ) {
                Text(
                    text = "$count",
                    style = TextStyle(
                        fontSize = 11.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = foreground,
                    ),
                )
            }
        }
    }
}

File: ios/Sources/SIT/Views/Shared/Warm/WarmFilterChip.swift
import SwiftUI

/// Pill-shaped filter chip used on the Network screen.
/// iOS uses `border-radius: 999` per HANDOFF lines 60-62.
struct WarmFilterChip: View {
    enum Kind {
        case all
        case category(WarmCategory)
        /// A user-created (non-canonical) group filter (TIC-88). Carries the
        /// group's emoji so the chip can show it where a category shows its SF
        /// Symbol. Styled neutrally (like `.all`) so custom-group chips read as
        /// distinct from the canonical category chips that follow "All".
        case userGroup(emoji: String)
    }

    let kind: Kind
    let label: String
    var count: Int? = nil
    let isActive: Bool
    let action: () -> Void

    var warmth: Warmth = .subtle

    var body: some View {
        Button(action: action) {
            HStack(spacing: 6) {
                if case .category(let cat) = kind {
                    Image(systemName: cat.systemImageName)
                        .font(.system(size: 11, weight: .semibold))
                }
                if case .userGroup(let emoji) = kind {
                    Text(emoji)
                        .font(.system(size: 12))
                }
                Text(label)
                    .font(.system(size: 13, weight: .semibold))
                    .lineLimit(1)
                    .fixedSize(horizontal: true, vertical: false)
                if let count {
                    Text("\(count)")
                        .font(.system(size: 11, weight: .semibold))
                        .padding(.horizontal, 6)
                        .padding(.vertical, 2)
                        .background(countBadgeBackground)
                        .foregroundStyle(countBadgeForeground)
                        .clipShape(Capsule())
                }
            }
            .padding(.horizontal, 14)
            .padding(.vertical, 10)
            .foregroundStyle(foreground)
            .background(background)
            .clipShape(Capsule())
            .overlay(
                Capsule().stroke(borderColor, lineWidth: 1)
            )
        }
        .buttonStyle(.plain)
    }

    private var palette: WarmPalette { WarmTheme.palette(for: warmth) }

    private var background: Color {
        switch (kind, isActive) {
        case (.all, true), (.userGroup, true):    palette.ink
        case (.all, false), (.userGroup, false):  palette.cardBg
        case (.category(let c), true):            c.palette.accent
        case (.category(let c), false):           c.palette.accentTint
        }
    }

    private var foreground: Color {
        switch (kind, isActive) {
        case (.all, true), (.userGroup, true):    palette.paper
        case (.all, false), (.userGroup, false):  palette.ink
        case (.category, true):                   Color(red: 0.98, green: 0.96, blue: 0.89)
        case (.category(let c), false):           c.palette.accent
        }
    }

    private var borderColor: Color {
        switch (kind, isActive) {
        case (.category, false), (.userGroup, false):  palette.cardBorder
        default:                                       Color.clear
        }
    }

    private var countBadgeBackground: Color {
        switch (kind, isActive) {
        case (.all, true), (.userGroup, true):    palette.paper.opacity(0.18)
        case (.all, false), (.userGroup, false):  palette.ink.opacity(0.06)
        case (.category, true):                   Color.white.opacity(0.22)
        case (.category(let c), false):           c.palette.accentBadge
        }
    }

    private var countBadgeForeground: Color {
        foreground
    }
}

#Preview("Inactive") {
    HStack(spacing: 8) {
        WarmFilterChip(kind: .all, label: "All", count: 142, isActive: false) {}
        WarmFilterChip(kind: .category(.family), label: "Family", count: 12, isActive: false) {}
        WarmFilterChip(kind: .category(.friends), label: "Friends", count: 24, isActive: false) {}
    }
    .padding()
    .background(WarmTheme.subtle.paper)
}

#Preview("Active") {
    HStack(spacing: 8) {
        WarmFilterChip(kind: .all, label: "All", count: 142, isActive: true) {}
        WarmFilterChip(kind: .category(.family), label: "Family", count: 12, isActive: true) {}
        WarmFilterChip(kind: .category(.work), label: "Work", count: 8, isActive: true) {}
    }
    .padding()
    .background(WarmTheme.subtle.paper)
}
